orc/src/main/java/org/apache/iceberg/orc/ORC.java

```diff
@@ -223,6 +232,161 @@ public DataWriteBuilder withSortOrder(SortOrder newSortOrder) {
     }
   }
 
+  public static DeleteWriteBuilder writeDeletes(OutputFile file) {
+    return new DeleteWriteBuilder(file);
+  }
+
+  public static class DeleteWriteBuilder {
+    private final WriteBuilder appenderBuilder;
+    private final String location;
+    private BiFunction<Schema, TypeDescription, OrcRowWriter<?>> createWriterFunc = null;
+    private Schema rowSchema = null;
+    private PartitionSpec spec = null;
+    private StructLike partition = null;
+    private EncryptionKeyMetadata keyMetadata = null;
+    private int[] equalityFieldIds = null;
+    private SortOrder sortOrder;
+    private Function<CharSequence, ?> pathTransformFunc = Function.identity();
```

Comment:
This seems to be never used. I think it is needed for query engine integrations to transform `CharSequence` to internal string representations. Should probably be used in positional delete?

Reply:
Can we use the introduced delete writers in `BaseWriterFactory` and adapt `TestWriterFactory` too? Then we will make sure deletes are working as expected.

We have Flink and Spark writer factories right now. Should be trivial to consume the new ORC writers there.

Reply:
> This seems to be never used. I think it is needed for query engine integrations to transform `CharSequence` to internal string representations. Should probably be used in positional delete?

I was debating whether to use this in ORC or not, decided to remove it until we have a better picture but some part accidentally remained in the code.
My thinking was:
- This function is there for `Parquet.java`, but only used when store the row data and `createWriterFunc` is not null. (maybe it is a bug?), also it is strange for me that we use `posDeleteSchema` only if the row schema is provided **and** we have a createWriterFunc as well. 
- This one is absolutely missing for `Avro.java`.

Maybe @openinx has more info about the reasons.

Reply:
> Can we use the introduced delete writers in `BaseWriterFactory` and adapt `TestWriterFactory` too? Then we will make sure deletes are working as expected.
> 
> We have Flink and Spark writer factories right now. Should be trivial to consume the new ORC writers there.

Will take a look how does it complicate / increase the size of the patch.

In the past I made the mistake to create huge PRs which made the job of the reviewer a hell, so here I tried to provide a first minimal patch.

Reply:
I have spent some time on this, and now I understand why the  `PositionDeleteStructWriter` different from the ORC part. Currently I think that it will take some serious work to make this working with Spark/Flink/Generic. Basically Spark has it's own implementation for ORC which is more-or-less independent of Generic.

I am inclined to separate out that work to a different PR (Disclaimer: I will be offline for 2 week starting this weekend, so either someone else picks this up, or it will take some time)

Reply:
The next step would be that we should add a new method to OrcRowWriter to access the internal ValueWriters:
```
List<OrcValueWriter<?>> writers();
```

This way we can create the `PositionDeleteStructWriter` for ORC using the internal writers.

Reply:
@szlta: Is working on enabling SparkOrcWriters for this usecase